src/main/java/reactor/ipc/netty/channel/ServerContextHandler.java

```diff
@@ -102,17 +115,59 @@ public boolean isDisposed() {
 		         .isActive();
 	}
 
+	private void createAndRegisterTerminateEmitterIfAbsent(Channel ch) {
+		DirectProcessor<Void> onTerminateEmitter = DirectProcessor.create();
+		if(ch.attr(ON_TERMINATE_EMITTER).compareAndSet(null, onTerminateEmitter)) {
+			FutureMono.from(ch.closeFuture()).subscribe(new CoreSubscriber<Void>() {
+				@Override
+				public void onSubscribe(Subscription s) {
+					if (!ch.attr(CLOSE_FUTURE_SUBSCRIPTION).compareAndSet(null, s)) {
+						throw new IllegalStateException(
+								"A previous subscription was present.");
+					}
+					onTerminateEmitter.onSubscribe(s);
+				}
+
+				@Override
+				public void onNext(Void aVoid) {
+					onTerminateEmitter.onNext(aVoid);
+				}
+
+				@Override
+				public void onError(Throwable t) {
+					onTerminateEmitter.onError(t);
+				}
+
+				@Override
+				public void onComplete() {
+					onTerminateEmitter.onComplete();
+				}
+			});
+		}
+	}
+
 	@Override
 	public void terminateChannel(Channel channel) {
-		if (!f.channel()
-		     .isActive()) {
+		completeAndDisposeTerminateEmitterIfPresent(channel);
+		if (!f.channel().isActive()) {
 			return;
 		}
-		if(!NettyContext.isPersistent(channel)){
+		if (!NettyContext.isPersistent(channel)) {
 			channel.close();
 		}
 	}
 
+	private void completeAndDisposeTerminateEmitterIfPresent(Channel channel) {
+		DirectProcessor<Void> terminateEmitter = channel.attr(ON_TERMINATE_EMITTER).getAndSet(null);
+		if (terminateEmitter != null) {
+			terminateEmitter.onComplete();
+		}
+		Subscription closeFutureSubscription = channel.attr(CLOSE_FUTURE_SUBSCRIPTION).getAndSet(null);
+		if (closeFutureSubscription != null) {
+			closeFutureSubscription.cancel();
+		}
+	}
+
```

Comment:
Just a note here - don't we need to `cancel` first the subscription and then to invoke `onComplete`on the emitter so that we are sure that we will not invoke `terminateEmitter.onComplete();` twice?

Reply:
double onComplete should be ignored by reactor operators. However yes it could make more sense to cancel first in that situation to avoid the later check.